bl/handlers.go

```diff
@@ -0,0 +1,212 @@
+package bl
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"github.com/go-redis/redis"
+	"github.com/olivere/elastic/v7"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"test/dal"
+)
+
+type db interface{
+	DBInitClient() error
+	DBGetBookByID(id string) (*elastic.GetResult,error)
+	DBGetStoreData() (int64, *elastic.SearchResult,error)
+	DBPutBook(book *dal.Book) (*elastic.IndexResponse,error)
+	DBUpdateBook(title string, id string) (*elastic.UpdateResponse,error)
+	DBDeleteBook(id string) (*elastic.DeleteResponse,error)
+	DBSearchBook(title string, authorName string, priceRange [2]float64, opcode int) (*elastic.SearchResult,error)
+}
+type userDataCache interface{
+	InitClient()
+	CacheAddActivity(username string, req string)error
+	CacheCreateNewClientActivity(username string, req string,client *redis.Client)
+	CacheUpdateClientActivity(username string, req string, client *redis.Client)
+	CacheGetUserActivity(username string) ([]interface{},error)
+}
+
+func GetBookByID(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	updateUserData(username,"GET",c.Request.RequestURI)
+	id := c.Param("id")
+	db := dal.Elastic{}
+	err := db.DBInitClient()
+	if err != nil{
+		handleError(err,c)
+	}
+	book, err := db.DBGetBookByID(id)
+	if err!=nil{
+		handleError(err,c)
+	}
+	c.JSON(http.StatusOK, book.Source)
+}
+
+func PutBook(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	updateUserData(username,"PUT",c.Request.RequestURI)
+	jsonData,err := ioutil.ReadAll(c.Request.Body)
+	if err!=nil{
+		handleError(err,c)
+	}
+	book := dal.Book{}
+	json.Unmarshal(jsonData,&book)
+	db := dal.Elastic{}
+	err = db.DBInitClient()
+	if err != nil{
+		handleError(err,c)
+	}
+	putStatus,err := db.DBPutBook(&book)
+	if err!=nil{
+		handleError(err,c)
+	}
+	c.JSON(http.StatusOK, putStatus.Id)
+}
+
+func UpdateBook(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	updateUserData(username,"POST",c.Request.RequestURI)
+	jsonData,err := ioutil.ReadAll(c.Request.Body)
+	if err!=nil{
+		handleError(err,c)	}
+	book := dal.Book{}
+	json.Unmarshal(jsonData,&book)
+	db := dal.Elastic{}
+	err = db.DBInitClient()
+	if err != nil{
+		handleError(err,c)
+	}
+	response,err := db.DBUpdateBook(book.Title,c.Param("id"))
```

Comment:
```suggestion
	response,err := db.DBUpdateBook(book.Title, book .Id)
```

Reply:
my book struct has no id field.
but I made it more clear now/

Reply:
so, don't send `c.Param("id")` into the `DBUpdateBook`,  first extract it to a separate variable
```suggestion
        id := c.Param("id")
	response,err := db.DBUpdateBook(book.Title,id))
```
and - what's happened if the id is empty?